File: sinkhorn.jl
using Printf, Einsum, LinearAlgebra


a = [.5, .5]
b = [.5, .5]
M = [0. 1.; 1. 0.]
reg = 0.01


function sinkhorn(a,b,M,reg;numItermax=1000,
    stopThr=1e-9,verbose=false,log_euv=false)

    a = reshape(a,:,1)
    b = reshape(b,:,1)

    if length(a) == 0
        a = ones((size(M)[1],)) / size(M)[1]
    end
    if length(b) == 0
        b = ones((size(M)[2],)) / size(M)[2]
    end

    # init data
    Nini = length(a)
    Nfin = length(b)

    if length(size(b)) > 2
        nbb = size(b)[2]
    else
        nbb = 0
    end

    if log_euv
        log_dict = Dict("err" => [])
    end

    if nbb > 0
        u = ones((Nini, nbb)) / Nini
        v = ones((Nfin, nbb)) / Nfin
    else
        u = ones(Nini) / Nini
        v = ones(Nfin) / Nfin
    end

    # K = Array{Float64}(undef, size(M))
    K = map(exp,-M ./ reg)

    tmp2 = Array{Float64}(undef, size(b))

    Kp = reshape(1 ./ a, :, 1) .* K
    cpt = 0
    err = 1
    while (err > stopThr && cpt < numItermax)
        uprev = u
        vprev = v

        KtransposeU = K' * u
        v = b ./ KtransposeU
        u = 1 ./ (Kp * v)

        if (any(x->x==0,KtransposeU) ||
            any(isnan,u) || any(isnan,v) ||
            any(isinf,u) || any(isinf,v))

            @printf("Warning: numerical errors at iteration %.0f",cpt)
            u = uprev
            v = vprev
            break
        end
        if cpt % 10 == 0
            #check every ten steps faster
            if nbb > 0
                err = sum((u-uprev)^2) / sum(u^2) + \
                sum((v-vprev)^2) / sum(v^2)
            else
                @einsum tmp2[j] = u[i]*M[i,j]*v[j]
                err = norm(tmp2 - b)^2
            end
            if log_euv
                push!(log_dict["err"],err)
            end

            if verbose
                if cpt % 200 == 0
                    @printf("%.5s|%.12s\n","It.","Err")
                    println("-"^19)
                    @printf("%.5d|%.8e",cpt,err)
                end
            end
        end
        cpt += 1
    end
    if log_euv
        log_dict = merge(log_dict,Dict("u"=>u),Dict("v"=>v))
    end

    if length(nbb) > 1
        res = Array{Float64}(undef, size(v)[2])
        @einsum res[k] = u[i,k]*K[i,j]*v[j,k]*M[i,j]
        if log_euv
            return res, log_dict
        else
            return res
        end
    else #return OT matrix

        if log_euv
            return reshape(u,:,1) .* K .* reshape(v,:,1), log_dict
        else
            return reshape(u,:,1) .* K .* reshape(v,:,1)
        end
    end
end

 @time sinkhorn(a,b,M,0.01)

File: sinkhorn.py
import numpy as np


def sinkhorn_knopp(a, b, M, reg, numItermax=1000,
                   stopThr=1e-9, verbose=False, log=False, **kwargs):

    a = np.asarray(a, dtype=np.float64)
    b = np.asarray(b, dtype=np.float64)
    M = np.asarray(M, dtype=np.float64)

    if len(a) == 0:
        a = np.ones((M.shape[0],), dtype=np.float64) / M.shape[0]
    if len(b) == 0:
        b = np.ones((M.shape[1],), dtype=np.float64) / M.shape[1]

    # init data
    Nini = len(a)
    Nfin = len(b)

    if len(b.shape) > 1:
        nbb = b.shape[1]
    else:
        nbb = 0

    if log:
        log = {'err': []}

    # we assume that no distances are null except those of the diagonal of
    # distances
    if nbb:
        u = np.ones((Nini, nbb)) / Nini
        v = np.ones((Nfin, nbb)) / Nfin
    else:
        u = np.ones(Nini) / Nini
        v = np.ones(Nfin) / Nfin

    # print(reg)

    # Next 3 lines equivalent to K= np.exp(-M/reg), but faster to compute
    K = np.empty(M.shape, dtype=M.dtype)
    np.divide(M, -reg, out=K)
    np.exp(K, out=K)

    # print(np.min(K))
    tmp2 = np.empty(b.shape, dtype=M.dtype)

    Kp = (1 / a).reshape(-1, 1) * K
    cpt = 0
    err = 1
    while (err > stopThr and cpt < numItermax):
        uprev = u
        vprev = v

        KtransposeU = np.dot(K.T, u)
        v = np.divide(b, KtransposeU)
        u = 1. / np.dot(Kp, v)

        if (np.any(KtransposeU == 0) or
                np.any(np.isnan(u)) or np.any(np.isnan(v)) or
                np.any(np.isinf(u)) or np.any(np.isinf(v))):
            # we have reached the machine precision
            # come back to previous solution and quit loop
            print('Warning: numerical errors at iteration', cpt)
            u = uprev
            v = vprev
            break
        if cpt % 10 == 0:
            # we can speed up the process by checking for the error only all
            # the 10th iterations
            if nbb:
                err = np.sum((u - uprev)**2) / np.sum((u)**2) + \
                    np.sum((v - vprev)**2) / np.sum((v)**2)
            else:
                # compute right marginal tmp2= (diag(u)Kdiag(v))^T1
                np.einsum('i,ij,j->j', u, K, v, out=tmp2)
                err = np.linalg.norm(tmp2 - b)**2  # violation of marginal
            if log:
                log['err'].append(err)

            if verbose:
                if cpt % 200 == 0:
                    print(
                        '{:5s}|{:12s}'.format('It.', 'Err') + '\n' + '-' * 19)
                print('{:5d}|{:8e}|'.format(cpt, err))
        cpt = cpt + 1
    if log:
        log['u'] = u
        log['v'] = v

    if nbb:  # return only loss
        res = np.einsum('ik,ij,jk,ij->k', u, K, v, M)
        if log:
            return res, log
        else:
            return res

    else:  # return OT matrix

        if log:
            return u.reshape((-1, 1)) * K * v.reshape((1, -1)), log
        else:
            return u.reshape((-1, 1)) * K * v.reshape((1, -1))
